File: src/com/wang/Game2dEngine/action/MoveCentreYToAction.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action;

import com.wang.Game2dEngine.sprite.Sprite;
import static java.lang.Math.abs;

/**
 *
 * @author ricolwang
 */
public class MoveCentreYToAction extends MoveAction
{

    MoveYToAction theMoveYToAction;

    double moveCentreYTo;
    float moveCentreYToDuration;

    public MoveCentreYToAction(Sprite theSprite)
    {
        if (theSprite != null)
        {
            this.setSprite(theSprite);

            this.moveCentreYTo = this.theSprite.getCentreY();
            this.moveCentreYToDuration = 0;

        } else
        {
            bComplete = true;
        }
    }

    public void MoveCentreYTo(double x, float duration)
    {
        if (duration <= 0)
        {
            duration = 0;
        }
        this.moveCentreYTo = x;
        this.moveCentreYToDuration = abs(duration * 1000);
    }

    @Override
    public void perform(double runningTime)
    {
        if (bComplete)
        {
            return;
        }

        if (theMoveYToAction == null)
        {
            theMoveYToAction = new MoveYToAction(this.theSprite);
            theMoveYToAction.moveYTo(moveCentreYTo - this.theSprite.getWidth() / 2.0, this.moveCentreYToDuration / 1000.0f);
            theMoveYToAction.setSprite(theSprite);
        }

        theMoveYToAction.perform(runningTime);
        if (theMoveYToAction.bComplete)
        {
            bComplete = true;
        }
    }

    @Override
    public void clearSprite()
    {
        this.theMoveYToAction.clearSprite();
        this.theMoveYToAction = null;
        this.theSprite = null;
    }
}

File: src/com/wang/Game2dEngine/action/MoveCentreYToAction.kt
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package com.wang.Game2dEngine.action

import com.wang.Game2dEngine.sprite.Sprite

/**
 * @author ricolwang
 */
class MoveCentreYToAction(theSprite: Sprite?) : MoveAction()
{
    var theMoveYToAction: MoveYToAction? = null
    var moveCentreYTo = 0.0
    var moveCentreYToDuration = 0f
    fun moveCentreYTo(y: Double, duration: Float)
    {
        var duration = duration
        if (duration <= 0)
        {
            duration = 0f
        }
        moveCentreYTo = y
        moveCentreYToDuration = Math.abs(duration * 1000)
    }

    override fun perform(runningTime: Double)
    {
        if (bComplete)
        {
            return
        }
        if (theMoveYToAction == null)
        {
            theMoveYToAction = MoveYToAction(theSprite)
            theMoveYToAction!!.moveYTo(moveCentreYTo - theSprite!!.width / 2.0, moveCentreYToDuration / 1000.0f)
            theMoveYToAction!!.setSprite(theSprite)
        }
        theMoveYToAction!!.perform(runningTime)
        if (theMoveYToAction!!.bComplete)
        {
            bComplete = true
        }
    }

    override fun clearSprite()
    {
        theMoveYToAction!!.clearSprite()
        theMoveYToAction = null
        theSprite = null
    }

    init
    {
        if (theSprite != null)
        {
            setSprite(theSprite)
            moveCentreYTo = this.theSprite!!.centreY
            moveCentreYToDuration = 0f
        } else
        {
            bComplete = true
        }
    }
}